File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/py_to_cpp/BoyerMooreSearch.cpp
#include <vector>
#include <string>

class BoyerMooreSearch {
private:
    std::string text;
    std::string pattern;
    int textLen;
    int patLen;

    int match_in_pattern(char c) {
        if (patLen == 0) {
            return -1;
        }
        for (int i = patLen - 1; i >= 0; i--) {
            if (pattern[i] == c) {
                return i;
            }
        }
        return -1;
    }

    int mismatch_in_text(int currentPos) {
        if (patLen == 0) {
            return -1;
        }
        for (int i = patLen - 1; i >= 0; i--) {
            if (pattern[i] != text[currentPos + i]) {
                return currentPos + i;
            }
        }
        return -1;
    }

public:
    BoyerMooreSearch(const std::string& text, const std::string& pattern) : 
        text(text), pattern(pattern), textLen(text.length()), patLen(pattern.length()) {}

    std::vector<int> bad_character_heuristic() {
        std::vector<int> positions;
        int i = 0;
        while (i <= textLen - patLen) {
            int mismatch_index = mismatch_in_text(i);
            if (mismatch_index == -1) {
                positions.push_back(i);
                i++;
            } else {
                int match_index = match_in_pattern(text[mismatch_index]);
                i = mismatch_index - match_index;
            }
        }
        return positions;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_py/code_output/BoyerMooreSearch.py
class BoyerMooreSearch:
    def __init__(self, text, pattern):
        self.text = text
        self.pattern = pattern
        self.textLen = len(text)
        self.patLen = len(pattern)
    
    def matchInPattern(self, ch):
        for i in range(self.patLen-1, -1, -1):
            if self.pattern[i] == ch:
                return i
        return -1

    def mismatchInText(self, currentPos):
        for i in range(self.patLen-1, -1, -1):
            if self.text[currentPos + i] != self.pattern[i]:
                return currentPos + i
        return -1

    def badCharacterHeuristic(self):
        badCharHeuristic = {}
        for j in range(self.patLen):
            badCharHeuristic[self.pattern[j]] = j
        
        if self.patLen == 0:
            return list(range(0, self.textLen + 1))
        
        positions = []
        i = 0
        while i <= self.textLen - self.patLen:
            mismatchIndex = self.mismatchInText(i)
            if mismatchIndex == -1:
                positions.append(i)
                i += self.patLen
            else:
                mismatchChar = self.text[mismatchIndex]
                matchIndex = badCharHeuristic.get(mismatchChar, -1)
                if matchIndex >= 0:
                    shift = mismatchIndex - i - matchIndex
                    if shift < 1:
                        shift = 1
                    i += shift
                else:
                    shift = mismatchIndex - i + 1
                    i += shift
        
        return positions